coprocessor/fhevm-engine/gw-listener/src/gw_listener.rs

```diff
@@ -112,58 +147,111 @@ impl<P: Provider<Ethereum> + Clone + 'static, A: AwsS3Interface> GatewayListener
+
                 item = verify_proof_request.next() => {
                     let Some(item) = item else {
                         error!("Block stream closed");
                         return Err(anyhow::anyhow!("Block stream closed"));
                     };
-                    let (request, log) = item?;
-                    self.verify_proof_request(db_pool, &mut from_block, request, log).await?;
+                    let (request, _) = item?;
+                    self.verify_proof_request(db_pool, request).await?;
                 }
-                item = activate_key.next() => {
-                    let Some(item) = item else {
-                        error!("Block stream closed");
-                        return Err(anyhow::anyhow!("Block stream closed"));
-                    };
-                    let (request, _log) = item?;
-                    self.activate_key(db_pool, request, &self.aws_s3_client, host_chain_id).await?;
-                    info!("ActivateKey event successful");
+            }
+            // Reset sleep duration on successful iteration.
+            self.reset_sleep_duration(sleep_duration);
+        }
+        Ok(())
+    }
+
+    async fn run_get_logs(
+        &self,
+        db_pool: &Pool<Postgres>,
+        sleep_duration: &mut u64,
+    ) -> anyhow::Result<()> {
+        let mut ticker = tokio::time::interval(self.conf.get_logs_poll_interval);
+        let mut last_processed_block_num = self.get_last_processed_block_num(db_pool).await?;
+
+        loop {
+            tokio::select! {
+                biased;
+
+                _ = self.cancel_token.cancelled() => {
+                    break;
                 }
-                item = activate_crs.next() => {
-                    let Some(item) = item else {
-                        error!("Block stream closed");
-                        return Err(anyhow::anyhow!("Block stream closed"));
+
+                _ = ticker.tick() => {
+                    let current_block = self.provider.get_block_number().await?;
+
+                    let from_block = if let Some(last) = last_processed_block_num {
+                        if last >= current_block {
+                            continue;
+                        }
+                        last + 1
+                    } else {
+                        current_block
+                    };
+
+                    let to_block = {
+                        let max = from_block.saturating_add(self.conf.get_logs_block_batch_size.saturating_sub(1));
+                        std::cmp::min(max, current_block)
                     };
-                    let (request, _log) = item?;
-                    self.activate_crs(db_pool, request, &self.aws_s3_client, host_chain_id).await?;
-                    info!("ActivateCrs event successful");
+
+                    let filter = Filter::new()
+                        .address(self.kms_generation_address)
+                        .from_block(from_block)
```

Comment:
the range could be to big here, we can cap here to X blocks, eg to_block.min(from_block + max_block_range_size)